# COVID-19 PANDEMIC

### Discussion Topic: COVID-19 pandemic; Advances in epidemiological modelling and risk communication.

##  Data Preparation
This is the process of organizing raw data to a format that can be used for analysis


In [966]:
#Importing relevant libraries
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns 


In [967]:
# Extracting the covid.zip
import zipfile
import os

# Specify the path to your zip file
zip_file_path = r'C:\Users\USER PC\Documents\Practisw\COVID-19\Covid-19-Pandemic\covid.zip'
# Specify the name of the file inside the zip archive
csv_file_name = 'COVID-19 Coronavirus.csv'

# Read the CSV file directly from the zip file
with zipfile.ZipFile(zip_file_path, 'r') as z:
    with z.open(csv_file_name) as f:
        df = pd.read_csv(f)

# Now df contains the data from the CSV file
df.head(10)

Country          Other names ISO 3166-1 alpha-3 CODE  \
0          Afghanistan          Afghanistan                     AFG   
1              Albania              Albania                     ALB   
2              Algeria              Algeria                     DZA   
3              Andorra              Andorra                     AND   
4               Angola               Angola                     AGO   
5             Anguilla             Anguilla                     AIA   
6  Antigua and Barbuda  Antigua and Barbuda                     ATG   
7            Argentina            Argentina                     ARG   
8              Armenia              Armenia                     ARM   
9                Aruba                Aruba                     ABW   

   Population                        Continent  Total Cases  Total Deaths  \
0    40462186                             Asia       177827          7671   
1     2872296                           Europe       273870          3492   
2    45236699                           Africa       265691          6874   
3       77481                           Europe        40024           153   
4    34654212                           Africa        99194          1900   
5       15237  Latin America and the Caribbean         2700             9   
6       99348  Latin America and the Caribbean         7493           135   
7    45921761  Latin America and the Caribbean      9041124        128065   
8     2972939                             Asia       422574          8617   
9      107560  Latin America and the Caribbean        34051           212   

   Tot Cases//1M pop  Tot Deaths/1M pop  Death percentage  
0               4395                190          4.313743  
1              95349               1216          1.275058  
2               5873                152          2.587216  
3             516565               1975          0.382271  
4               2862                 55          1.915438  
5             177200                591          0.333333  
6              75422               1359          1.801682  
7             196881               2789          1.416472  
8             142140               2898          2.039169  
9             316577               1971          0.622596

In [968]:
# Repeat the same process for the remaining zip files
import zipfile
import os

# Specify the path to your zip file
zip_file_path = r'C:\Users\USER PC\Documents\Practisw\COVID-19\Covid-19-Pandemic\covid2.zip'
# Specify the name of the file inside the zip archive
csv_file_name = 'Covid Live.csv'

# Read the CSV file directly from the zip file
with zipfile.ZipFile(zip_file_path, 'r') as a:
    with a.open(csv_file_name) as b:
        df2 = pd.read_csv(b)
        
        
# Zipfile2
import zipfile
import os

# Specify the path to your zip file
zip_file_path = r'C:\Users\USER PC\Documents\Practisw\COVID-19\Covid-19-Pandemic\cov3.zip'
# Specify the name of the file inside the zip archive
csv_file_name = 'covid_19_clean_complete new.csv'

# Read the CSV file directly from the zip file
with zipfile.ZipFile(zip_file_path, 'r') as p:
    with p.open(csv_file_name) as g:
        df3 = pd.read_csv(g)
        
        


In [969]:
df2.head(10)

# Country,\nOther Total\nCases Total\nDeaths  New\nDeaths  \
0   1             USA   98,166,904     1,084,282          NaN   
1   2           India   44,587,307       528,629          NaN   
2   3          France   35,342,950       155,078          NaN   
3   4          Brazil   34,706,757       686,027          NaN   
4   5         Germany   33,312,373       149,948          NaN   
5   6        S. Korea   24,769,101        28,406         42.0   
6   7              UK   23,621,952       189,919          NaN   
7   8           Italy   22,432,803       177,054          NaN   
8   9           Japan   21,229,216        44,678          NaN   
9  10          Russia   20,948,470       387,163          NaN   

  Total\nRecovered Active\nCases Serious,\nCritical Tot Cases/\n1M pop  \
0       94,962,112     2,120,510              2,970            293,206   
1       44,019,095        39,583                698             31,698   
2       34,527,115       660,757                869            538,892   
3       33,838,636       182,094              8,318            161,162   
4       32,315,200       847,225              1,406            397,126   
5       24,013,461       727,234                352            482,547   
6       23,347,871        84,162                146            344,857   
7       21,779,604       476,145                141            372,250   
8       20,380,089       804,449                204            169,043   
9       19,979,763       581,544              2,300            143,674   

  Deaths/\n1M pop   Total\nTests Tests/\n1M pop     Population  
0           3,239  1,118,158,870      3,339,729    334,805,269  
1             376    894,416,853        635,857  1,406,631,776  
2           2,365    271,490,188      4,139,547     65,584,518  
3           3,186     63,776,166        296,146    215,353,593  
4           1,788    122,332,384      1,458,359     83,883,596  
5             553     15,804,065        307,892     51,329,899  
6           2,773    522,526,476      7,628,357     68,497,907  
7           2,938    246,636,495      4,092,684     60,262,770  
8             356     76,201,407        606,772    125,584,838  
9           2,655    273,400,000      1,875,095    145,805,947

In [970]:
df3.head(10)

Country/Region        Lat        Long        Date  Confirmed  Deaths  \
0          Afghanistan  3.393.911  67.709.953  2020-01-22          0       0   
1              Albania    411.533     201.683  2020-01-22          0       0   
2              Algeria    280.339      1.6596  2020-01-22          0       0   
3              Andorra    425.063      1.5218  2020-01-22          0       0   
4               Angola   -112.027     178.739  2020-01-22          0       0   
5  Antigua and Barbuda    170.608    -617.964  2020-01-22          0       0   
6            Argentina   -384.161    -636.167  2020-01-22          0       0   
7              Armenia    400.691     450.382  2020-01-22          0       0   
8            Australia   -354.735   1.490.124  2020-01-22          0       0   
9            Australia   -338.688   1.512.093  2020-01-22          0       0   

   Recovered  Active             WHO Region  
0          0       0  Eastern Mediterranean  
1          0       0                 Europe  
2          0       0                 Africa  
3          0       0                 Europe  
4          0       0                 Africa  
5          0       0               Americas  
6          0       0               Americas  
7          0       0                 Europe  
8          0       0        Western Pacific  
9          0       0        Western Pacific

In [971]:
import zipfile
import os

# Specify the path to your zip file
zip_file_path = r'C:\Users\USER PC\Documents\Practisw\COVID-19\Covid-19-Pandemic\archive (10).zip'
# Specify the name of the file inside the zip archive
csv_file_name = 'open_one_time_covid_education_impact.csv'

# Read the CSV file directly from the zip file
with zipfile.ZipFile(zip_file_path, 'r') as z:
    with z.open(csv_file_name) as f:
        df4 = pd.read_csv(f)
        
df4.head()

submission_id submission_date  gender                 age  \
0  4570534489882624      2021-03-17  Female   Over 45 years old   
1  6440166989496320      2021-03-29    Male  26 to 35 years old   
2  5001167808757760      2021-03-18  Female  26 to 35 years old   
3  5524933164924928      2021-03-24    Male  36 to 45 years old   
4  5032515751641088      2021-03-18  Female   Over 45 years old   

                          geography  \
0               Suburban/Peri-urban   
1               Suburban/Peri-urban   
2  City center or metropolitan area   
3               Suburban/Peri-urban   
4               Suburban/Peri-urban   

                                 financial_situation  \
0  I can afford food and regular expenses, but no...   
1          I cannot afford enough food for my family   
2  I can comfortably afford food, clothes, and fu...   
3                I can afford food, but nothing else   
4  I can afford food and regular expenses, but no...   

                                education  \
0  University or college degree completed   
1  University or college degree completed   
2  University or college degree completed   
3  University or college degree completed   
4  University or college degree completed   

                                   employment_status submission_state  \
0                                    I am unemployed          Miranda   
1                                    I am unemployed          Miranda   
2  I work full-time, either as an employee or sel...          Miranda   
3  I work full-time, either as an employee or sel...          Miranda   
4                                    I am unemployed          Miranda   

   are_there_children_0_to_2_yrs_out_of_educational_system  ...  \
0                                                  0        ...   
1                                                  0        ...   
2                                                  1        ...   
3                                                  0        ...   
4                                                  0        ...   

   are_children_3_to_17_yrs_dealing_with_irregular_school_activity  \
0                                                  0                 
1                                                  1                 
2                                                  1                 
3                                                  1                 
4                                                  1                 

   are_children_being_teached_by_unqualified_people  \
0                                                 0   
1                                                 0   
2                                                 1   
3                                                 1   
4                                                 0   

   did_teachers_leave_the_educational_system  \
0                                          0   
1                                          1   
2                                          1   
3                                          1   
4                                          1   

   do_school_and_the_teachers_have_internet_connection  \
0                                                  1     
1                                                  0     
2                                                  0     
3                                                  0     
4                                                  0     

   do_children_have_internet_connection  \
0                                     1   
1                                     1   
2                                     1   
3                                     1   
4                                     1   

  do_children_3_to_17_yrs_miss_virtual_class_due_to_lack_of_electricity  \
0                                                  0                      
1                                                  1                      
2                                           

In [972]:
# create new dataframes
df5 = pd.read_csv('populations_states.csv')
df5.head()

date data_or_inferred  incarcerated_total  incarcerated_white  \
0  2000-01-01         raw data             25213.0                8753   
1  2000-02-01         raw data             25213.0                8753   
2  2000-03-01         raw data             25161.0                8723   
3  2000-04-01         raw data             25200.0                8738   
4  2000-05-01         raw data             25226.0                8788   

   incarcerated_black  incarcerated_hispanic  incarcerated_amerind  \
0             16382.0                    NaN                   NaN   
1             16382.0                    NaN                   NaN   
2             16361.0                    NaN                   NaN   
3             16380.0                    NaN                   NaN   
4             16358.0                    NaN                   NaN   

   incarcerated_asian  incarcerated_other  incarcerated_unknown  \
0                 NaN                78.0                   NaN   
1                 NaN                78.0                   NaN   
2                 NaN                77.0                   NaN   
3                 NaN                82.0                   NaN   
4                 NaN                80.0                   NaN   

   incarcerated_nonwhite    state                                     url  
0                16460.0  Alabama  http://www.doc.state.al.us/StatReports  
1                16460.0  Alabama  http://www.doc.state.al.us/StatReports  
2                16438.0  Alabama  http://www.doc.state.al.us/StatReports  
3                16462.0  Alabama  http://www.doc.state.al.us/StatReports  
4                16438.0  Alabama  http://www.doc.state.al.us/StatReports

In [973]:
df6 = pd.read_csv('admissions_releases_states.csv')
df6.head()

date  admissions_total  admissions_white  admissions_black  \
0  2015-02-01              1583               600               193   
1  2015-03-01              1580               615               189   
2  2015-04-01              1779               706               235   
3  2015-05-01              1821               730               260   
4  2015-06-01              1689               704               215   

   admissions_hispanic  admissions_amerind  admissions_asian  \
0                633.0               130.0               6.0   
1                640.0               114.0               3.0   
2                701.0               115.0               4.0   
3                669.0               125.0               8.0   
4                618.0               129.0               0.0   

   admissions_other  releases_total  releases_white  releases_black  \
0              21.0            1649             653             194   
1              19.0            1559             644             169   
2              18.0            1693             703             206   
3              29.0            1659             685             193   
4              23.0            1570             651             174   

   releases_hispanic  releases_amerind  releases_asian  releases_other  \
0              688.0              85.0             7.0            22.0   
1              610.0             111.0             6.0            19.0   
2              657.0              93.0             5.0            29.0   
3              616.0             136.0             3.0            26.0   
4              603.0             111.0             7.0            24.0   

     state  
0  Arizona  
1  Arizona  
2  Arizona  
3  Arizona  
4  Arizona

# Exploratory Data Analysis

## Renaming columns
In our dataframes exists columns with same name or meaning but some column names seem to be modified e.g Country and Country/Region all mean the same

In [974]:
# Renaming columns in df3b
df3 = df3.rename(columns={'Country/Region': 'Country', 'Deaths': 'Total Deaths'})
df3 = df3.reset_index(drop=True)
df3

Country                     Lat        Long        Date  \
0                Afghanistan               3.393.911  67.709.953  2020-01-22   
1                    Albania                 411.533     201.683  2020-01-22   
2                    Algeria                 280.339      1.6596  2020-01-22   
3                    Andorra                 425.063      1.5218  2020-01-22   
4                     Angola                -112.027     178.739  2020-01-22   
...                      ...                     ...         ...         ...   
49063  Sao Tome and Principe                   1.864      6.6131  2020-07-27   
49064                  Yemen  15.552.726.999.999.900  48.516.388  2020-07-27   
49065                Comoros                -116.455     433.333  2020-07-27   
49066             Tajikistan                  38.861     712.761  2020-07-27   
49067                Lesotho                  -29.61     282.336  2020-07-27   

       Confirmed  Total Deaths  Recovered  Active             WHO Region  
0              0             0          0       0  Eastern Mediterranean  
1              0             0          0       0                 Europe  
2              0             0          0       0                 Africa  
3              0             0          0       0                 Europe  
4              0             0          0       0                 Africa  
...          ...           ...        ...     ...                    ...  
49063        865            14        734     117                 Africa  
49064       1691           483        833     375  Eastern Mediterranean  
49065        354             7        328      19                 Africa  
49066       7235            60       6028    1147                 Europe  
49067        505            12        128     365                 Africa  

[49068 rows x 9 columns]

In [975]:
# Rename columns in df2
df2 = df2.rename(columns={'Country,\nOther': 'Country', 'Total\nCases': 'Total Cases', 'Total\nTests': 'Total Tests', 'Total\nDeaths': 'Total Deaths' })
df2.head()

#  Country Total Cases Total Deaths  New\nDeaths Total\nRecovered  \
0  1      USA  98,166,904    1,084,282          NaN       94,962,112   
1  2    India  44,587,307      528,629          NaN       44,019,095   
2  3   France  35,342,950      155,078          NaN       34,527,115   
3  4   Brazil  34,706,757      686,027          NaN       33,838,636   
4  5  Germany  33,312,373      149,948          NaN       32,315,200   

  Active\nCases Serious,\nCritical Tot Cases/\n1M pop Deaths/\n1M pop  \
0     2,120,510              2,970            293,206           3,239   
1        39,583                698             31,698             376   
2       660,757                869            538,892           2,365   
3       182,094              8,318            161,162           3,186   
4       847,225              1,406            397,126           1,788   

     Total Tests Tests/\n1M pop     Population  
0  1,118,158,870      3,339,729    334,805,269  
1    894,416,853        635,857  1,406,631,776  
2    271,490,188      4,139,547     65,584,518  
3     63,776,166        296,146    215,353,593  
4    122,332,384      1,458,359     83,883,596

In [976]:
df5 = df5.rename(columns={'incarcerated_white': 'Hospital Capacity', 'date': 'Date', 'incarcerated_white': 'Lockdown Restrictions'})

In [977]:
df4 = df4.rename(columns={'submission_date': 'Date'})

In [978]:
df6 = df6.rename(columns={'admissions_black': 'Vaccination Rate', 'releases_black': 'Hospital Capacity'})
df6.head()

date  admissions_total  admissions_white  Vaccination Rate  \
0  2015-02-01              1583               600               193   
1  2015-03-01              1580               615               189   
2  2015-04-01              1779               706               235   
3  2015-05-01              1821               730               260   
4  2015-06-01              1689               704               215   

   admissions_hispanic  admissions_amerind  admissions_asian  \
0                633.0               130.0               6.0   
1                640.0               114.0               3.0   
2                701.0               115.0               4.0   
3                669.0               125.0               8.0   
4                618.0               129.0               0.0   

   admissions_other  releases_total  releases_white  Hospital Capacity  \
0              21.0            1649             653                194   
1              19.0            1559             644                169   
2              18.0            1693             703                206   
3              29.0            1659             685                193   
4              23.0            1570             651                174   

   releases_hispanic  releases_amerind  releases_asian  releases_other  \
0              688.0              85.0             7.0            22.0   
1              610.0             111.0             6.0            19.0   
2              657.0              93.0             5.0            29.0   
3              616.0             136.0             3.0            26.0   
4              603.0             111.0             7.0            24.0   

     state  
0  Arizona  
1  Arizona  
2  Arizona  
3  Arizona  
4  Arizona

In [979]:
df6.head()

date  admissions_total  admissions_white  Vaccination Rate  \
0  2015-02-01              1583               600               193   
1  2015-03-01              1580               615               189   
2  2015-04-01              1779               706               235   
3  2015-05-01              1821               730               260   
4  2015-06-01              1689               704               215   

   admissions_hispanic  admissions_amerind  admissions_asian  \
0                633.0               130.0               6.0   
1                640.0               114.0               3.0   
2                701.0               115.0               4.0   
3                669.0               125.0               8.0   
4                618.0               129.0               0.0   

   admissions_other  releases_total  releases_white  Hospital Capacity  \
0              21.0            1649             653                194   
1              19.0            1559             644                169   
2              18.0            1693             703                206   
3              29.0            1659             685                193   
4              23.0            1570             651                174   

   releases_hispanic  releases_amerind  releases_asian  releases_other  \
0              688.0              85.0             7.0            22.0   
1              610.0             111.0             6.0            19.0   
2              657.0              93.0             5.0            29.0   
3              616.0             136.0             3.0            26.0   
4              603.0             111.0             7.0            24.0   

     state  
0  Arizona  
1  Arizona  
2  Arizona  
3  Arizona  
4  Arizona

## Dropping unnecessary columns

In [980]:
# Dropping columns in df2
df2 = df2.drop(['Population', 'New\nDeaths',
       'Total\nRecovered', 'Active\nCases', 'Serious,\nCritical',
       'Tot Cases/\n1M pop', 'Deaths/\n1M pop',
       'Tests/\n1M pop'], axis=1)
df2.head()

#  Country Total Cases Total Deaths    Total Tests
0  1      USA  98,166,904    1,084,282  1,118,158,870
1  2    India  44,587,307      528,629    894,416,853
2  3   France  35,342,950      155,078    271,490,188
3  4   Brazil  34,706,757      686,027     63,776,166
4  5  Germany  33,312,373      149,948    122,332,384

In [981]:
# df3 dropping columns
df3 = df3.drop(['Confirmed','Recovered', 'Active', 'WHO Region', 'Total Deaths' ], axis=1)
df3

Country                     Lat        Long        Date
0                Afghanistan               3.393.911  67.709.953  2020-01-22
1                    Albania                 411.533     201.683  2020-01-22
2                    Algeria                 280.339      1.6596  2020-01-22
3                    Andorra                 425.063      1.5218  2020-01-22
4                     Angola                -112.027     178.739  2020-01-22
...                      ...                     ...         ...         ...
49063  Sao Tome and Principe                   1.864      6.6131  2020-07-27
49064                  Yemen  15.552.726.999.999.900  48.516.388  2020-07-27
49065                Comoros                -116.455     433.333  2020-07-27
49066             Tajikistan                  38.861     712.761  2020-07-27
49067                Lesotho                  -29.61     282.336  2020-07-27

[49068 rows x 4 columns]

In [982]:
df5

Date data_or_inferred  incarcerated_total  Lockdown Restrictions  \
0     2000-01-01         raw data             25213.0                   8753   
1     2000-02-01         raw data             25213.0                   8753   
2     2000-03-01         raw data             25161.0                   8723   
3     2000-04-01         raw data             25200.0                   8738   
4     2000-05-01         raw data             25226.0                   8788   
...          ...              ...                 ...                    ...   
9580  2022-05-01         raw data            156704.0                  90396   
9581  2022-06-01         raw data            157426.0                  90785   
9582  2022-07-01         raw data            157435.0                  90792   
9583  2022-08-01         raw data            157565.0                  90837   
9584  2022-09-01         raw data            158041.0                  91122   

      incarcerated_black  incarcerated_hispanic  incarcerated_amerind  \
0                16382.0                    NaN                   NaN   
1                16382.0                    NaN                   NaN   
2                16361.0                    NaN                   NaN   
3                16380.0                    NaN                   NaN   
4                16358.0                    NaN                   NaN   
...                  ...                    ...                   ...   
9580             60041.0                    NaN                3998.0   
9581             60370.0                    NaN                3997.0   
9582             60386.0                    NaN                3995.0   
9583             60470.0                    NaN                4015.0   
9584             60652.0                    NaN                4023.0   

      incarcerated_asian  incarcerated_other  incarcerated_unknown  \
0                    NaN                78.0                   NaN   
1                    NaN                78.0                   NaN   
2                    NaN                77.0                   NaN   
3                    NaN                82.0                   NaN   
4                    NaN                80.0                   NaN   
...                  ...                 ...                   ...   
9580              2269.0                 NaN                   NaN   
9581              2274.0                 NaN                   NaN   
9582              2262.0                 NaN                   NaN   
9583              2243.0                 NaN                   NaN   
9584              2244.0                 NaN                   NaN   

      incarcerated_nonwhite    state  \
0                   16460.0  Alabama   
1                   16460.0  Alabama   
2                   16438.0  Alabama   
3                   16462.0  Alabama   
4                   16438.0  Alabama   
...                     ...      ...   
9580                66308.0  Federal   
9581                66641.0  Federal   
9582                66643.0  Federal   
9583                66728.0  Federal   
9584                66919.0  Federal   

                                                    url  
0                http://www.doc.state.al.us/StatReports  
1                http://www.doc.state.al.us/StatReports  
2                http://www.doc.state.al.us/StatReports  
3                http://www.doc.state.al.us/StatReports  
4                http://www.doc.state.al.us/StatReports  
...                                                 ...  
9580  https://www.bop.gov/about/statistics/statistic...  
9581  https://www.bop.gov/about/statistics/statistic...  
9582  https://www.bop.gov/about/statistics/statistic...  
9583  https://www.bop.gov/about/statistics/statistic...  
9584  https://www.bop.gov/about/statistics/statistic...  

[9585 rows x 13 columns]

In [983]:
# dropping multiple columns in df4
df4 = df4.drop(['Date', 'financial_situation', 'education', 'employment_status',
       'submission_state',
       'are_there_children_0_to_2_yrs_out_of_educational_system',
       'were_children_3_to_17_yrs_enrolled_and_did_not_return_to_school',
       'are_there_children_who_stopped_enrolling_in_primary_education',
       'are_there_children_who_stopped_enrolling_in_secondary_education',
       'are_children_attending_face_to_face_classes',
       'can_children_observe_deterioration_of_basic_services_of_school',
       'do_children_3_and_17_yrs_receive_regular_school_meals',
       'are_there_teachers_at_scheduled_class_hours',
       'are_children_3_to_17_yrs_dealing_with_irregular_school_activity',
       'are_children_being_teached_by_unqualified_people',
       'did_teachers_leave_the_educational_system',
       'do_school_and_the_teachers_have_internet_connection',
       'do_children_have_internet_connection',
       'do_children_3_to_17_yrs_miss_virtual_class_due_to_lack_of_electricity',
       'does_home_shows_severe_deficit_of_electricity',
       'does_home_shows_severe_deficit_of_internet',
       'do_children_3_to_17_yrs_miss_class_or_in_lower_grade',
       'are_children_promoted_with_a_modality_different_from_formal_evaluation', 'geography'], axis=1)


In [984]:
df4.head()

submission_id  gender                 age
0  4570534489882624  Female   Over 45 years old
1  6440166989496320    Male  26 to 35 years old
2  5001167808757760  Female  26 to 35 years old
3  5524933164924928    Male  36 to 45 years old
4  5032515751641088  Female   Over 45 years old

In [985]:
df4.shape

(4436, 3)

In [986]:

# Keep only the specified columns
df = df[['Country', 'Total Cases', 'Total Deaths']]

# Display the updated DataFrame
df.head()

Country  Total Cases  Total Deaths
0  Afghanistan       177827          7671
1      Albania       273870          3492
2      Algeria       265691          6874
3      Andorra        40024           153
4       Angola        99194          1900

In [987]:
df6 = df6.drop(['releases_hispanic', 'releases_amerind',
       'releases_asian', 'releases_other', 'state', 'admissions_total', 'admissions_white', 'admissions_hispanic', 'admissions_amerind', 'admissions_asian',
       'admissions_other', 'releases_total', 'releases_white', 'date'], axis=1)
df6.head()

Vaccination Rate  Hospital Capacity
0               193                194
1               189                169
2               235                206
3               260                193
4               215                174

In [988]:
df5 = df5.drop([ 'data_or_inferred', 'incarcerated_total', 'incarcerated_black', 'incarcerated_hispanic', 'incarcerated_amerind','incarcerated_asian', 'incarcerated_other', 'incarcerated_unknown','incarcerated_nonwhite', 'state', 'url', 'Date'], axis=1)

In [989]:
n = 5000
df5 = df5.head(5000)

In [990]:
df.head()

Country  Total Cases  Total Deaths
0  Afghanistan       177827          7671
1      Albania       273870          3492
2      Algeria       265691          6874
3      Andorra        40024           153
4       Angola        99194          1900

## Filling Missing Values
We will fill missing values by imputation. Imputation is a technique used to fill in missing values in a dataset. Here are some common techniques.

 Mean/Median Imputation:

Concept: Replace missing values with the mean or median of the corresponding column.

In [991]:
import pandas as pd

# Assuming your dataframe is called "df"
df5['Lockdown Restrictions'].fillna(df5['Lockdown Restrictions'].mean(), inplace=True)


## Merging our dataframes
We do this by joining the dataframes with similar columns, then dropping unrelevant columns in our analysis

In [992]:
merge1 = pd.merge(df, df3, on='Country', how='outer')
merge1.head()

Country  Total Cases  Total Deaths        Lat        Long        Date
0  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-22
1  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-23
2  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-24
3  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-25
4  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-26

### Check for missing values

In [993]:
merge1.isna().sum()

Country            0
Total Cases     4700
Total Deaths    4700
Lat               53
Long              53
Date              53
dtype: int64

In [994]:
# fill missing values
merge1[['Lat', 'Long', 'Date']] = merge1[['Lat', 'Long', 'Date']].bfill()

In [995]:
# confirm all missing values are solved
print(merge1.isna().sum())

Country            0
Total Cases     4700
Total Deaths    4700
Lat                0
Long               0
Date               0
dtype: int64


In [996]:
# Sample DataFrame
# Reduce to the first 5000 rows
n = 5000
merge1 = merge1.head(5000)

In [997]:
df5.isna().sum()

Lockdown Restrictions    0
dtype: int64

In [998]:
merge2 = pd.concat([merge1, df5], axis=1)
merge2.head()

Country  Total Cases  Total Deaths        Lat        Long        Date  \
0  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-22   
1  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-23   
2  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-24   
3  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-25   
4  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-26   

   Lockdown Restrictions  
0                   8753  
1                   8753  
2                   8723  
3                   8738  
4                   8788

In [999]:
#check for missing values
merge2.isna().sum()

Country                  0
Total Cases              0
Total Deaths             0
Lat                      0
Long                     0
Date                     0
Lockdown Restrictions    0
dtype: int64

## Manipulate the number of rows
 I want to expand our dataframe df6 to have atleast 5000 rows.

In [1000]:
import pandas as pd

# Assuming df6 is your original DataFrame
# Calculate how many times we need to repeat the DataFrame
repeat_times = (5000 // len(df6)) + 1  # +1 to ensure we have enough rows

# Repeat the DataFrame and then slice it to get exactly 5000 rows
expanded_df = pd.concat([df6] * repeat_times, ignore_index=True).head(5000)

# Check the shape of the new DataFrame
print(expanded_df.shape)

(5000, 2)


In [1001]:
expanded_df.head()

Vaccination Rate  Hospital Capacity
0               193                194
1               189                169
2               235                206
3               260                193
4               215                174

In [1002]:
expanded_df.isna().sum()

Vaccination Rate     0
Hospital Capacity    0
dtype: int64

In [1003]:
# Merging the 3rd dataframe
merge3 = pd.concat([merge2, expanded_df], axis=1)
merge3.head()

Country  Total Cases  Total Deaths        Lat        Long        Date  \
0  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-22   
1  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-23   
2  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-24   
3  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-25   
4  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-26   

   Lockdown Restrictions  Vaccination Rate  Hospital Capacity  
0                   8753               193                194  
1                   8753               189                169  
2                   8723               235                206  
3                   8738               260                193  
4                   8788               215                174

In [1004]:
df4.head()

submission_id  gender                 age
0  4570534489882624  Female   Over 45 years old
1  6440166989496320    Male  26 to 35 years old
2  5001167808757760  Female  26 to 35 years old
3  5524933164924928    Male  36 to 45 years old
4  5032515751641088  Female   Over 45 years old

In [1005]:
df5.head()

Lockdown Restrictions
0                   8753
1                   8753
2                   8723
3                   8738
4                   8788

In [1006]:
df4.shape


(4436, 3)

In [1007]:
df4.head()

submission_id  gender                 age
0  4570534489882624  Female   Over 45 years old
1  6440166989496320    Male  26 to 35 years old
2  5001167808757760  Female  26 to 35 years old
3  5524933164924928    Male  36 to 45 years old
4  5032515751641088  Female   Over 45 years old

In [1008]:
# Expand the dataframe df4 to fit the merged columns

import pandas as pd

# Assuming df6 is your original DataFrame
# Calculate how many times we need to repeat the DataFrame
repeat_times2 = (5000 // len(df4)) + 1  # +1 to ensure we have enough rows

# Repeat the DataFrame and then slice it to get exactly 5000 rows
new_df4 = pd.concat([df4] * repeat_times, ignore_index=True).head(5000)

# Check the shape of the new DataFrame
print(new_df4.shape)


(5000, 3)


In [1009]:
new_df4.head()

submission_id  gender                 age
0  4570534489882624  Female   Over 45 years old
1  6440166989496320    Male  26 to 35 years old
2  5001167808757760  Female  26 to 35 years old
3  5524933164924928    Male  36 to 45 years old
4  5032515751641088  Female   Over 45 years old

In [1010]:
# Concatenate the last dataframes 
merged_df = pd.concat([merge3, new_df4], axis=1)
merged_df.head()

Country  Total Cases  Total Deaths        Lat        Long        Date  \
0  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-22   
1  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-23   
2  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-24   
3  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-25   
4  Afghanistan     177827.0        7671.0  3.393.911  67.709.953  2020-01-26   

   Lockdown Restrictions  Vaccination Rate  Hospital Capacity  \
0                   8753               193                194   
1                   8753               189                169   
2                   8723               235                206   
3                   8738               260                193   
4                   8788               215                174   

      submission_id  gender                 age  
0  4570534489882624  Female   Over 45 years old  
1  6440166989496320    Male  26 to 35 years old  
2  5001167808757760  Female  26 to 35 years old  
3  5524933164924928    Male  36 to 45 years old  
4  5032515751641088  Female   Over 45 years old

In [1012]:
# Confirm if our final dataframe is clean
merged_df.isna().sum()

Country                  0
Total Cases              0
Total Deaths             0
Lat                      0
Long                     0
Date                     0
Lockdown Restrictions    0
Vaccination Rate         0
Hospital Capacity        0
submission_id            0
gender                   0
age                      0
dtype: int64

## Dropping Duplicates 

In [1013]:
# Check for duplicates
duplicates = merged_df.duplicated()
print(duplicates)

0       False
1       False
2       False
3       False
4       False
        ...  
4995    False
4996    False
4997    False
4998    False
4999    False
Length: 5000, dtype: bool


### Check for correlation of columns

In [1018]:
# Select only numeric columns
numeric_df = merged_df.select_dtypes(include=['number'])

# Calculate the correlation matrix
correlation_matrix = numeric_df.corr()

correlation_matrix.head()

Total Cases  Total Deaths  Lockdown Restrictions  \
Total Cases               1.000000      0.618891               0.068698   
Total Deaths              0.618891      1.000000              -0.145168   
Lockdown Restrictions     0.068698     -0.145168               1.000000   
Vaccination Rate          0.015601      0.012013              -0.035419   
Hospital Capacity         0.016350      0.006576              -0.029702   

                       Vaccination Rate  Hospital Capacity  submission_id  
Total Cases                    0.015601           0.016350      -0.007733  
Total Deaths                   0.012013           0.006576      -0.027027  
Lockdown Restrictions         -0.035419          -0.029702      -0.006773  
Vaccination Rate               1.000000           0.942477      -0.033746  
Hospital Capacity              0.942477           1.000000      -0.033747

## Correlation Coefficients Interpretation

### Total Cases and Total Deaths (0.618891):
There is a moderate positive correlation between total cases and total deaths. This suggests that as the number of total cases increases, the number of total deaths tends to increase as well. This is expected in a pandemic context, where more cases can lead to more deaths.

### Total Cases and Lockdown Restrictions (0.068698):
The correlation is very weak and positive. This indicates that there is almost no relationship between the total number of cases and the severity of lockdown restrictions. This could suggest that lockdown measures may not have been effective in reducing the number of cases, or that other factors are at play.

### Total Cases and Vaccination Rate (0.015601):
The correlation is very close to zero, indicating no significant relationship between total cases and vaccination rates in this dataset. This could imply that the vaccination rate did not have a noticeable impact on the total number of cases during the period analyzed.

### Total Cases and Hospital Capacity (0.016350):
Again, this correlation is very weak, suggesting that there is no significant relationship between total cases and hospital capacity. This could indicate that hospital capacity was not a limiting factor in the number of cases reported.

### Total Deaths and Lockdown Restrictions (-0.145168):
There is a weak negative correlation between total deaths and lockdown restrictions. This suggests that as lockdown restrictions become more severe, the total number of deaths may slightly decrease. However, the correlation is weak, so this relationship is not strong.

### Total Deaths and Vaccination Rate (0.012013):
This correlation is very close to zero, indicating no significant relationship between total deaths and vaccination rates in this dataset.

### Total Deaths and Hospital Capacity (0.006576):
This correlation is also very weak, suggesting no significant relationship between total deaths and hospital capacity.

### Lockdown Restrictions and Vaccination Rate (-0.035419):
There is a very weak negative correlation, indicating that there is almost no relationship between lockdown restrictions and vaccination rates.

### Lockdown Restrictions and Hospital Capacity (-0.029702):
This correlation is also very weak, suggesting no significant relationship between lockdown restrictions and hospital capacity.

### Vaccination Rate and Hospital Capacity (0.942477):
There is a very strong positive correlation between vaccination rate and hospital capacity. This suggests that as the vaccination rate increases, hospital capacity also tends to increase. This could imply that higher vaccination rates are associated with better hospital preparedness or capacity.

## Data Visualization